### PR TITLE
feat(transformer): add `object-spread` plugin

### PR DESCRIPTION
**Pull request by magic-akari**

### CODE REVIEW

**Comment by magic-akari** [comment]
* **#3133** <a href="https://app.graphite.dev/github/pr/oxc-project/oxc/3133?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `main`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @magic-akari and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Comment by magic-akari** [comment]
`module_imports.rs` may need to be improved to enable testing, but I do not want to make changes to it in this PR.

**Comment by magic-akari** [comment]
> Hi @magic-akari. Just wondered how complete this is and whether we should try to finish it off?

It's up to you; I'm not in a rush.

> Transformer has changed a bit since April - it now uses `Traverse` rather than `VisitMut` - but I doubt updating it for that change would be difficult, as they're largely the same.

Wow, I am very willing to learn new methods.

**Comment by magic-akari** [comment]
I am currently refactoring this pull request.

**Comment by overlookmotel** [comment]
Hi @magic-akari. Sorry I didn't reply before. My Github notifications is like a snowdrift!

In case you've not noticed it, we recently adopted a "style guide" for writing transformers: https://github.com/oxc-project/oxc/blob/main/crates/oxc_transformer/README.md

Most of the existing transformers have not yet been brought in line with the style guide, but [this one](https://github.com/oxc-project/oxc/blob/main/crates/oxc_transformer/src/es2020/nullish_coalescing_operator.rs) has, so it's a good model to follow.

**Comment by magic-akari** [comment]
```bash
❯ cargo run -p oxc_transformer --example transformer tasks/coverage/babel/packages/babel-parser/test/fixtures/es2018/object-rest-spread/5/input.js
    Finished `dev` profile [unoptimized] target(s) in 0.15s
     Running `target/debug/examples/transformer tasks/coverage/babel/packages/babel-parser/test/fixtures/es2018/object-rest-spread/5/input.js`
Original:

z = {x, ...y}

Transformed:

import { default as _objectSpread } from "@babel/helpers/lib/helpers/objectSpread2.js";
z = _objectSpread({ x }, y);
```

File: crates/oxc_transformer/src/options/transformer.rs
Comment (Dunqing) [inline crates/oxc_transformer/src/options/transformer.rs:185]: Nitpick: Can you put the `es2018` above `es2019`

File: crates/oxc_transformer/src/es2018/options.rs
Comment (Dunqing) [inline crates/oxc_transformer/src/es2018/options.rs:11]: Should be deleted?

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (Dunqing) [inline crates/oxc_transformer/src/es2018/object_spread.rs:1]: We need to add some comments here.  You can refer to #4972

**Review by Dunqing — CHANGES_REQUESTED** [review]

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/object_spread.rs:20]: Babel also handles the transformation for `rest` expressions, but this feature is not included in this pull request. Moreover, there is a issue with Babel's handling of `rest` parameters; I recommend emulating the behavior of esbuild. See https://github.com/swc-project/swc/issues/8527

File: crates/oxc_transformer/src/es2018/options.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/options.rs:11]: The `object_rest` has not been implemented yet. How to mark TODO in OXC?

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (Dunqing) [reply crates/oxc_transformer/src/es2018/object_spread.rs:48]: Although doesn't support the `rest` expression yet, we still should use `ObjectRestSpread` as a plugin name. Because we need to transform both `rest` and `spread` in the same plugin. We should always follow `Babel` and [Proposal](https://github.com/tc39/proposal-object-rest-spread) as possible

> Moreover, there is a issue with Babel's handling of rest expressions; I recommend emulating the behavior of esbuild. See https://github.com/swc-project/swc/issues/8527

Is possible to fix the issue in Babel's implementation? We should follow Babel's implementation as much as possible for now. This way we can reuse `Babel's` tests and easier debugging of errors caused by porting wrong

Does this make sense to you?

**Comment by Dunqing** [comment]
Oh, this plugin used `@babel/helpers`, we may hold this until #4753 has a conclusion.

**Comment by overlookmotel** [comment]
If the helper is the only outstanding thing on this PR, I suggest we merge it now and open a separate issue to remind us to go back and update the helper mechanism once we have an answer on #4753. I expect we'll be making changes to `Traverse` APIs over next week or so, so this PR will likely accrue merge conflicts if we leave it open for long.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/object_spread.rs:40]: Done.

**Review by Dunqing — DISMISSED** [review]
Have conflicting files need to be resolved

**Comment by Dunqing** [comment]
I just found that don't have any test changes related to this plugin. You need to uncomment this line
https://github.com/oxc-project/oxc/blob/6ad403b0ce8829a41aca06e452f225fb64d08de8/tasks/transform_conformance/src/constants.rs#L27

**Comment by magic-akari** [comment]
> I just found that don't have any test changes related to this plugin. You need to uncomment this line
> 
> https://github.com/oxc-project/oxc/blob/6ad403b0ce8829a41aca06e452f225fb64d08de8/tasks/transform_conformance/src/constants.rs#L27


https://github.com/babel/babel/blob/f42e24d6451d333052900e1ae073e4bc5d484cca/packages/babel-plugin-transform-object-rest-spread/test/fixtures/assumption-pureGetters/spread-single-call/output.js#L1-L7

In Babel's test cases, `babelHelpers.objectSpread2` is used instead of the imported helpers. I should make a special return when running the tests. For example,

https://github.com/oxc-project/oxc/blob/91e2652c4ff7e37b3c6c0802a0110eb6c42dd134/crates/oxc_transformer/src/es2018/object_spread.rs#L138-L159

```diff
         Expression::from(ctx.ast.member_expression_static(SPAN, object, property, false))
     }
 
+    #[cfg(feature = "external_helper")]
+    fn babel_external_helper(ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
+        let ident_object =
+            ctx.create_reference_id(SPAN, ctx.ast.atom("babelHelpers"), None, ReferenceFlags::Read);
+        let object = ctx.ast.expression_from_identifier_reference(ident_object);
+        let property = ctx.ast.identifier_name(SPAN, "_objectSpread");
+
+        Expression::from(ctx.ast.member_expression_static(SPAN, object, property, false))
+    }
+
     fn get_extend_object_callee(&mut self, ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
+        if cfg!(feature = "external_helper") {
+            return Self::babel_external_helper(ctx);
+        }
+
         if self.options.set_spread_properties {
             Self::object_assign(ctx)
         } else {
```

What do you think?

**Comment by overlookmotel** [comment]
The 3% perf drop is strange. NB: The benchmark time measured includes also running the parser and semantic, and transformer itself is only a fraction of the total time, therefore a negative perf result of this size is an underestimate.

If you look at the flamegraph on CodSpeed, you can see that on the `pdf.mjs` benchmark, the transformer itself gets 28% slower on this PR. That's quite a lot!

We may want to look into that before merging.

**Comment by Dunqing** [comment]
> https://github.com/babel/babel/blob/f42e24d6451d333052900e1ae073e4bc5d484cca/packages/babel-plugin-transform-object-rest-spread/test/fixtures/assumption-pureGetters/spread-single-call/output.js#L1-L7
> 
> In Babel's test cases, `babelHelpers.objectSpread2` is used instead of the imported helpers. I should make a special return when running the tests. For example,
> 
> https://github.com/oxc-project/oxc/blob/91e2652c4ff7e37b3c6c0802a0110eb6c42dd134/crates/oxc_transformer/src/es2018/object_spread.rs#L138-L159
> 
> ```diff
>          Expression::from(ctx.ast.member_expression_static(SPAN, object, property, false))
>      }
>  
> +    #[cfg(feature = "external_helper")]
> +    fn babel_external_helper(ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
> +        let ident_object =
> +            ctx.create_reference_id(SPAN, ctx.ast.atom("babelHelpers"), None, ReferenceFlags::Read);
> +        let object = ctx.ast.expression_from_identifier_reference(ident_object);
> +        let property = ctx.ast.identifier_name(SPAN, "_objectSpread");
> +
> +        Expression::from(ctx.ast.member_expression_static(SPAN, object, property, false))
> +    }
> +
>      fn get_extend_object_callee(&mut self, ctx: &mut TraverseCtx<'a>) -> Expression<'a> {
> +        if cfg!(feature = "external_helper") {
> +            return Self::babel_external_helper(ctx);
> +        }
> +
>          if self.options.set_spread_properties {
>              Self::object_assign(ctx)
>          } else {
> ```
> 
> What do you think?

We're not sure how we're going to solve #4753. So let's make it as simple as possible, so we'll just return `babel_external_helper`

File: crates/oxc_transformer/src/es2018/mod.rs
Comment (Dunqing) [inline crates/oxc_transformer/src/es2018/mod.rs:20]: Should be `Option<ObjectRestSpread<'a>>`, we need this to determine if we should transform

File: crates/oxc_transformer/src/es2018/mod.rs
Comment (Dunqing) [inline crates/oxc_transformer/src/es2018/mod.rs:38]: We only run it when this plugin is enabled. You can refer to https://github.com/oxc-project/oxc/blob/d304d6f9734b486f63eb829ab719b85f52d705c2/crates/oxc_transformer/src/es2020/mod.rs#L53-L57

```suggestion
        if self.object_rest_spread.is_some() {
          self.object_rest_spread.enter_expression(expr, ctx);
        }
```

I think that's why so many unrelated test changes were made

**Comment by overlookmotel** [comment]
Can I just check: Are we running the conformance tests from this Babel plugin? If so, they're all passing - in that case, bravo! With the introduction of the transform semantic checker, that's no mean feat.

**Comment by magic-akari** [comment]
> Can I just check: Are we running the conformance tests from this Babel plugin? If so, they're all passing - in that case, bravo! With the introduction of the transform semantic checker, that's no mean feat.

I'm not sure about the workings of conformance testing.
However, I think that achieving a 100% pass is something wrong.
This is because there's an unimplemented part of the plugin.
It implies that some parts of the test code may not be working as expected.

**Comment by overlookmotel** [comment]
> I'm not sure about the workings of conformance testing.

Me neither! @Dunqing Can you advise?

**Comment by magic-akari** [comment]
I've caught it. It's been additionally disabled here 
https://github.com/oxc-project/oxc/blob/3fa863525e60b67ecb981e9210191ab127f76a2c/tasks/transform_conformance/src/constants.rs#L66-L68

**Comment by overlookmotel** [comment]
OK great. Now how do we disable just the rest-related tests? Or maybe we don't? @Dunqing?

**Review by overlookmotel — COMMENTED** [review]
I've made a bunch of comments. I hope they're useful.

As I said in one of them, please don't feel the need to address everything immediately. The priority is to get the tests passing first.

File: crates/oxc_transformer/src/es2018/mod.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/mod.rs:14]: This piece of code appears at the top of every other transformer mod. I guess it's used to prevent lint warnings for ctx and options.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/object_spread.rs]: Here, a pre-order traversal is expected. But I think the order might not matter. I'll test if simply removing it leads to the correct result.
This might be the actual cause of the performance degradation.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/object_spread.rs]: If I've understood correctly, I should avoid using a user-defined `Object` and instead refer to the global `Object`, correct?
If the user has created their own `Object` variable, I should rename it, perhaps to something like `Object2`.

File: crates/oxc_transformer/src/lib.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/lib.rs]: This is the legacy code.
It was originally written in the object_spread module.
Considering that every use of babel helper will insert the import statement when exiting the program in the future, a unified insertion would be more efficient.

However, we are now using only `babelHelpers.objectSpread2` to pass the test and waiting for the decision on #4753, so this part is already legacy and unclean code. It's OK to just delete them.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/object_spread.rs:59]: I generally believe the performance here is acceptable, as it operates at O(n) complexity and will early return.
However, if we delegate the detection logic to another visitor, we'd have to manage the interdependencies between them.
Additionally, there's a risk that once `exit_spread_element` is exited, it could be mutated by another padd visitor, leading to the flag being out of sync with the actual state.

File: crates/oxc_transformer/src/es2018/mod.rs
Comment (overlookmotel) [reply crates/oxc_transformer/src/es2018/mod.rs:14]: Ah I see. Sorry.

I don't know why we're storing `Ctx` and options in lots of places, and not using them. Seems wasteful and leads to more `Rc` cloning. But OK, leave it as is. We'll go through all the transformers and clean up this anti-pattern later on.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (overlookmotel) [reply crates/oxc_transformer/src/es2018/object_spread.rs]: OK cool. I doubt this is where perf regression is coming from as this code is only reached if a `SpreadElement` is found, and the vast majority of object literals won't use spread.

As far as I can see, going in forwards order should work equally well (but I could of course be wrong). But I tend not to assume that how Babel does things is necessarily the "best way", as Babel has less emphasis on performance than Oxc does, and so doesn't always choose the most efficient algorithms. Our modus operandi tends to be: start by copying Babel, and then try to improve on it!

It's not essential to turn it into a loop, it's just preferable where it's possible, as loops tend to be faster.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (overlookmotel) [reply crates/oxc_transformer/src/es2018/object_spread.rs]: I think we can just accept that if the source code has a variable called `Object`, the transformed code is not going to work properly. I think that's OK - overriding globals is asking for trouble, so don't complain if you get it!

My point is not that we need to fix that, but that our scope tree should completely match the AST. So if we insert `Object.assign` and `Object` in that position resolves to a local variable, the scope tree should correctly reflect that.

PS: If we *really* wanted to tackle this, we could use `{}.constructor.assign` instead of `Object.assign` where there's an existing binding for `Object`. But even that's not infallible - what if the user overwrote `Object.prototype.constructor`? So I don't think it's worthwhile trying to handle. Ultimately you can't 100% prevent the user from sabotaging you.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (overlookmotel) [reply crates/oxc_transformer/src/es2018/object_spread.rs:59]: All of that is true - co-ordinating multiple visitors is tricky. But please note my comment above that the addition of this transform is slowing down the transformer on one benchmark by 28%. That's not negligible.

I may be wrong that these lines is where the perf loss is coming from, but it's my best guess.

File: crates/oxc_transformer/src/lib.rs
Comment (overlookmotel) [reply crates/oxc_transformer/src/lib.rs]: Yes, it's fine to leave it as is for now. I just need to think about a good scheme for sharing state between different transforms in general.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/object_spread.rs]: OK, that makes sense.

File: crates/oxc_transformer/src/es2018/object_spread.rs
Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/object_spread.rs:59]: I removed the `self.enter_expression(expr, ctx);` above and the CodSpeed gate passed. So I think the bottleneck is there, not here.

Comment (magic-akari) [reply crates/oxc_transformer/src/es2018/object_spread.rs:59]: The benchmark of pdf.mjs has returned to its original level.

image

Comment (overlookmotel) [reply crates/oxc_transformer/src/es2018/object_spread.rs:59]: It's really odd because `pdf.js` only has 4 x object spreads in it. So I can't explain how that could be enough to produce a 28% perf drop. But it seems it did. Great!

**Comment by Dunqing** [comment]
> OK great. Now how do we disable just the rest-related tests? Or maybe we don't? @Dunqing?

These disabled plugins are used in other plugins. i.e. when we develop the `transform-typescript` plugin, and the `transform-object-rest-spread` plugin is not supported yet, so we don't care about tests where `transform-typescript` and `transform-object-rest-spread` are both enabled.

Now, we can remove `transform-object-rest-spread` because we are doing support this plugin right now

**Comment by overlookmotel** [comment]
I'm not sure if I was clear. This PR adds support for the *spread* part of the `transform-object-rest-spread` but the *rest* part is left as a TODO. So should we:

1. Enable all the tests for this plugin, and all the ones related to rest will show as failing in snapshot?
2. Disable the tests related to rest, so only the spread-related tests (which should all pass) run?

**Comment by Dunqing** [comment]
> 1. Enable all the tests for this plugin, and the rest-related tests will show as failing in snapshot?

Yes 

> 2. Disable the rest-related tests, so only the spread-related tests run (and should all pass)?


This depends on whether the spread part is fully completed or not

**Comment by Dunqing** [comment]
This PR is now large, and there will always be conflicts. I think we should merge first and then iterate for improvement. @overlookmotel What do you think?

**Comment by overlookmotel** [comment]
> This PR is now large, and there will always be conflicts. I think we should merge first and then iterate for improvement.

Agreed. I've merged latest main. If the tests pass, I'll merge it.

**Comment by overlookmotel** [comment]
Thanks @magic-akari for your work on this. Sorry it's taken a while to get merged.

**Pull request merged**
